File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 
func preorderTraversal(root *TreeNode) []int {
    var stack []*TreeNode
    var result []int = []int{}

    current := root
    for current != nil || len(stack) > 0 {
        if current != nil {
            result = append(result, current.Val)
            if current.Right != nil {
                stack = append(stack, current.Right)
            }
            current = current.Left
        } else {
            current = stack[len(stack) - 1]
            stack = stack[:len(stack) - 1]
        }
    }

    return result
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        const result = this.dfs(root);
        return result !== -1;
    }

    dfs(root) {
        if (!root) {
            return 0;
        }

        const left = this.dfs(root.left);
        const right = this.dfs(root.right);

        if (left === -1 || right === -1) {
            return -1;
        }

        if (Math.abs(right - left) > 1) {
            return -1;
        }

        return 1 + Math.max(left, right);
    }
}
